src/main/test/java/io/burt/jmespath/contrib/function/JmesPathRuntimeWithStringFunctionTest.java

```diff
@@ -0,0 +1,586 @@
+package io.burt.jmespath.contrib.function;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+import java.util.regex.PatternSyntaxException;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import io.burt.jmespath.Adapter;
+import io.burt.jmespath.RuntimeConfiguration;
+import io.burt.jmespath.function.ArgumentTypeException;
+import io.burt.jmespath.function.FunctionRegistry;
+import io.burt.jmespath.parser.ParseException;
+
+public abstract class JmesPathRuntimeWithStringFunctionTest<T> extends JmesPathRuntimeTest<T> {
+  private FunctionRegistry functionRegistry = FunctionRegistry.defaultRegistry()
+          .extend(new ConcatFunction())
+          .extend(new LowerCaseFunction())
+          .extend(new MatchesFunction())
+          .extend(new NormalizeSpaceFunction())
+          .extend(new ReplaceFunction())
+          .extend(new SubstringAfterFunction())
+          .extend(new SubstringBeforeFunction())
+          .extend(new TokenizeFunction())
+          .extend(new TranslateFunction())
+          .extend(new UpperCaseFunction());
+
+  private Adapter<T> runtime = createRuntime(RuntimeConfiguration.builder()
+          .withFunctionRegistry(functionRegistry)
+          .build());
+
+  @Override
+  protected Adapter<T> runtime() { return runtime; }
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  private T dontCare, emptyObject, emptyList;
+
+  @Before
+  public void setUp() {
+    emptyObject = parse("{}");
+    emptyList = parse("[]");
+    dontCare = emptyObject;
+  }
+  /**
+   * The following test cases are inspired by the examples of the related functions from XPath Function Specification:
+   * https://www.w3.org/TR/xpath-functions-31/
+   * Those examples just illustrate the statements of specification rules and are considered as non-normative.
+   */
```

Comment:
This is a comment that I think better belongs in a commit message. Should it remain if someone adds more tests not inspired by the XPath spec? It's good information, but I don't think adding it as a comment is beneficial in the long run. Having it in the commit message would be, though, because I can see exactly which tests were meant by "the following test cases" (i.e. those committed in the commit that included this message).